src/schema/sources.ts

```diff
@@ -347,56 +375,131 @@ const sourceToGQL = (source: Source): GQLSource => ({
 });
 
 export enum SourcePermissions {
-  View,
-  Post,
-  Leave,
-  Delete,
-  Edit,
+  View = 'view',
+  Post = 'post',
+  PostDelete = 'post_delete',
+  RemoveMember = 'remove_member',
+  Leave = 'leave',
+  Delete = 'delete',
+  Edit = 'edit',
+}
+
+const canRemoveMemberRoles = [
+  SourceMemberRoles.Owner,
+  SourceMemberRoles.Moderator,
+];
+
+const canPostDeleteRoles = [
+  SourceMemberRoles.Owner,
+  SourceMemberRoles.Moderator,
+];
+
+export const hasGreaterAccessCheck = (
+  loggedUser: SourceMember,
+  member: SourceMember,
+) => {
+  const memberRank = roleRank[member.role];
+  const loggedUserRank = roleRank[loggedUser.role];
+  const hasGreaterAccess = loggedUserRank > memberRank;
+
+  if (!hasGreaterAccess) {
+    throw new ForbiddenError('Access denied!');
+  }
+};
+
+interface SourcePermissionsProps {
+  permission: SourcePermissions;
+  validateRankAgainstId: string;
 }
 
+const hasPermissionCheck = (
+  source: Source | GQLSource,
+  member: SourceMember | GQLSourceMember,
+  permission: SourcePermissions,
+) => {
+  switch (permission) {
+    case SourcePermissions.Post:
+      if (source.type !== SourceType.Squad) {
+        return false;
+      }
+      break;
+    case SourcePermissions.PostDelete:
+      if (!canPostDeleteRoles.includes(member.role)) {
+        return false;
+      }
+      break;
+    case SourcePermissions.RemoveMember:
+      if (!canRemoveMemberRoles.includes(member.role)) {
```

Comment:
Just a thought should we maybe not include `hasGreaterAccessCheck` here?
Then all access happens in one switch?

Reply:
Good point. We can probably do that, so all checks will happen in a single place.